scapy/contrib/tcpros.py

```diff
@@ -0,0 +1,732 @@
+
+
+class TCPROSElement(Packet):
+    """
+    Captures each one of the elements in the
+    TCPROSHeader or TCPROSBody packages.
+
+    NOTE: Used within other packages
+    """
+
+    name = "TCPROSElement"
+    fields_desc = [
+        # field
+        FieldLenField(name="field_length",
+                      default=None, length_of="field", fmt="<I"),
+        StrLenField(name="field", length_from=lambda pkt: pkt.field_length,
+                    default=""),
+    ]
+
+    def extract_padding(self, s):
+        return "", s
+
+
+class TCPROSHeader(Packet):
+    """
+    The Header of the TCPROS package.
+
+        0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+        +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+        |header.|  len1 |                    element1                   |
+        +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+        |               |  len2 |                                       |
+        +-+-+-+-+-+-+-+-+-+-+-+-+                                       +
+        |                            element2                           |
+        +                       +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+        |                       |  ...  |                               |
+        +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+                               +
+        |                                                               |
+        +                              ...                              +
+        |                                                               |
+        +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+
+    Generated with:
+        protocol "header_length:4,len1:4,element1:32,len2:4,
+            element2:64,...:4,...:80" --bits 32
+
+    Typical header looks like:
+
+        7C 00 00 00 12 00 00 00 63 61 6C 6C 65 72 69 64  |.......callerid
+        3D 2F 6C 69 73 74 65 6E 65 72 27 00 00 00 6D 64  =/listener'...md
+        35 73 75 6D 3D 39 39 32 63 65 38 61 31 36 38 37  5sum=992ce8a1687
+        63 65 63 38 63 38 62 64 38 38 33 65 63 37 33 63  cec8c8bd883ec73c
+        61 34 31 64 31 0D 00 00 00 74 63 70 5F 6E 6F 64  a41d1....tcp_nod
+        65 6C 61 79 3D 30 0E 00 00 00 74 6F 70 69 63 3D  elay=0....topic=
+        2F 63 68 61 74 74 65 72 14 00 00 00 74 79 70 65  /chatter....type
+        3D 73 74 64 5F 6D 73 67 73 2F 53 74 72 69 6E 67  =std_msgs/String
+
+    """
+
+    name = "TCPROSHeader"
+    fields_desc = [
+        # header_length
+        FieldLenField(name="header_length",
+                      default=None, length_of="list", fmt="<I"),
+        # list  ## contains TCPROSElement
+        PacketListField("list", None,
+                        TCPROSElement, count_from=lambda pkt: nfields),
+    ]
+
+    def pre_dissect(self, s):
+        """
+        Called to prepare the layer before dissection
+        """
+        # Obtain the number of fields in the header
+        global nfields
```

Comment:
I wouldn't recommend a global variable here. If you need to store this value, you could add a slot to your packet class. 
```
class TCPROSHeader(Packet):
        __slots__ = Packet.__slots__ + ["nfields"]
...
```
This would be more cleaner. 
in your fields_desc you could ask for it like this:
```
... count_from=lambda pkt: pkt.nfields ...
```

Reply:
see 904af5e528d4ab2c3bf9eb9d75e3e7a933e8a471